src/recon_buildblock/RelativeDifferencePrior.cxx

```diff
@@ -430,6 +512,121 @@ add_multiplication_with_approximate_Hessian(DiscretisedDensity<3,elemT>& output,
+    error("LogcoshPrior: kappa image has not the same index range as the reconstructed image\n");
+
+  const int min_z = output.get_min_index();
+  const int max_z = output.get_max_index();
+  for (int z=min_z; z<=max_z; z++)
+  {
+    const int min_dz = max(weights.get_min_index(), min_z-z);
+    const int max_dz = min(weights.get_max_index(), max_z-z);
+
+    const int min_y = output[z].get_min_index();
+    const int max_y = output[z].get_max_index();
+
+    for (int y=min_y;y<= max_y;y++)
+    {
+      const int min_dy = max(weights[0].get_min_index(), min_y-y);
+      const int max_dy = min(weights[0].get_max_index(), max_y-y);
+
+      const int min_x = output[z][y].get_min_index();
+      const int max_x = output[z][y].get_max_index();
+
+      for (int x=min_x;x<= max_x;x++)
+      {
+        const int min_dx = max(weights[0][0].get_min_index(), min_x-x);
+        const int max_dx = min(weights[0][0].get_max_index(), max_x-x);
+
+        /// At this point, we have j = [z][y][x]
+        // The next for loops will have k = [z+dz][y+dy][x+dx]
+        // The following computes
+        //(H_{wf} y)_j =
+        //      \sum_{k\in N_j} w_{(j,k)} f''_{d}(x_j,x_k) y_j +
+        //      \sum_{(i \in N_j) \ne j} w_{(j,i)} f''_{od}(x_j, x_i) y_i
+        // Note the condition in the second sum that i is not equal to j
+
+        elemT result = 0;
+        for (int dz=min_dz;dz<=max_dz;++dz)
+          for (int dy=min_dy;dy<=max_dy;++dy)
+            for (int dx=min_dx;dx<=max_dx;++dx)
+            {
+              elemT current = weights[dz][dy][dx];
+              if (dz == dy == dz == 0) {
+                // The j == k case
+                current *= diagonal_second_derivative(current_estimate[z][y][x],
+                                                      current_estimate[z + dz][y + dy][x + dx]) * input[z][y][x];
+              } else {
+                current *= (diagonal_second_derivative(current_estimate[z][y][x],
+                                                       current_estimate[z + dz][y + dy][x + dx]) * input[z][y][x] +
+                            off_diagonal_second_derivative(current_estimate[z][y][x],
+                                                           current_estimate[z + dz][y + dy][x + dx]) *
+                            input[z + dz][y + dy][x + dx]);
+              }
+
+              if (do_kappa)
+                current *= (*kappa_ptr)[z][y][x] * (*kappa_ptr)[z+dz][y+dy][x+dx];
+
+              result += current;
+            }
+
+        output[z][y][x] += result * this->penalisation_factor;
+      }
+    }
+  }
+  return Succeeded::yes;
+}
+
+template <typename elemT>
+float
+RelativeDifferencePrior<elemT>::
+diagonal_second_derivative(const float x_j, const float x_k) const
+{
+  if (x_j > 0.0 || x_k > 0.0 || this->epsilon > 0.0)
+    return 2 * pow(2 * x_k + this->epsilon, 2) /
+           pow(x_j + x_k + this->gamma * abs(x_j - x_k) + this->epsilon, 3);
+  else
+    return 0.0;
```

Comment:
I guess not defined? Should document what we do here. Is 0 a reasonable value?

Reply:
An interesting point that was not considered at the time. If x_j, x_k and epsilon = 0, it would compute `(0^2) / (0^3) = undefined`. However, I think that 0 is the only value we return and retain numerical stability but I do not think there is a gerneral correct answer here. Hopefully by setting the value to 0 we can minimise any impact of this instance.

Reply:
probably right limit will be +inf, and left limit -inf then, which means the limit doesn't exist. Setting it to 0 is "reasonable" but should be documented.